### PR TITLE
add sync methods to sink

### PR DESCRIPTION
**Pull request by sveisvei**
## JIRA Issue
[COREWEB-39](https://jira.finn.no/browse/COREWEB-39)

## Status
**READY**

## Description
Adds async methods to support setting and getting files without streams, needed for simpler meta-data storage interface.

## Todos
- [x] Tests
- [x] Documentation

## Related PRs
* https://github.com/asset-pipe/asset-pipe-sink-mem/pull/46

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,57 +17,57 @@
...
 [![Greenkeeper badge](https://badges.greenkeeper.io/asset-pipe/asset-pipe-sink-gcs.svg)](https://greenkeeper.io/)
 
-A [asset-pipe][asset-pipe] sink for writing to and reading from [Google Cloud Storage][gcs].
-
-The intention of the [asset-pipe][asset-pipe] sink modules is to use be able to write and read files
-to different backends by just swapping modules. By each sink implementing the same public API it is
-possible to ex use this module in one environment and another sink module in another environment.
-
-These sinks are normally used by the [asset-pipe-build-server][asset-pipe-build-server].
+A [asset-pipe][asset-pipe] sink for writing to and reading from [Google Cloud
```
Comment (digitalsadhu) [inline README.md]: s/a/an

File: README.md [hunk 2]
```diff
@@ -17,57 +17,57 @@
...
+Storage][gcs].
 
+The intention of the [asset-pipe][asset-pipe] sink modules is to use be able to
```
Comment (digitalsadhu) [inline README.md]: ~~use~~

File: README.md [hunk 3]
```diff
@@ -17,57 +17,57 @@
...
+write and read files to different backends by just swapping modules. By each
+sink implementing the same public API it is possible to ex use this module in
```
Comment (digitalsadhu) [inline README.md]: s/By each sink implementing/Because each sink implements

Comment (digitalsadhu) [reply README.md]: ~~ex~~

File: README.md [hunk 4]
```diff
@@ -17,57 +17,57 @@
...
-
 ## Example
 
 Read an asset feed from [Google Cloud Storage][gcs] and serve it on http:
 
 ```js
-const express = require('express');
-const Sink = require('asset-pipe-sink-gcs');
+const express = require("express");
+const Sink = require("asset-pipe-sink-gcs");
```
Comment (digitalsadhu) [inline README.md]: looks like we should tweak the prettier markdown settings to use single quotes

File: README.md [hunk 5]
```diff
@@ -76,70 +76,81 @@ This module have the following API:
...
+  read files to/from - Required.
+* `metadata` - Object - A bucket [metadata object][gcs-bucket-meta] - Optional.
 
 Events:
 
- - `storage info` - General info on the underlaying storage. Emits with: `message` and `metaObject`.
+* `storage info` - General info on the underlaying storage. Emits with:
+  `message` and `metaObject`.
+
+### get(fileName)
```
Comment (digitalsadhu) [inline README.md]: Maybe should indicate return type:
What about just writing types or some such??

###` get(filename: string): Promise<string>`

File: README.md [hunk 6]
```diff
@@ -76,70 +76,81 @@ This module have the following API:
...
+* `type` - String - File type of the file to write. Used as extension of the
+  persisted file - Required.
 
 Events:
 
- - `file saved` - When a file have been sucessfully persisted. Emits with: `id` and `file`.
- - `file not saved` -  When a file could not be persisted. Emits with: `error`.
- - `error` -  When an error occured during persistence. Emits with: `error`.
-
+* `file saved` - When a file have been sucessfully persisted. Emits with: `id`
```
Comment (digitalsadhu) [inline README.md]: s/have/has

File: lib/sink.js [hunk 7]
```diff
@@ -73,61 +71,95 @@ class WriteStream extends stream.PassThrough {
...
+                    `Bucket "${bucketName}" did not exist. Created bucket`
+                );
+                return this.bucket;
+            })
+            .catch(error => {
+                this.bucket = this.gcs.bucket(bucketName);
+                this.emit(
                     'storage info',
-                    `Bucket "${bucket}" exist. Using bucket`,
+                    `Bucket "${bucketName}" exist. Using bucket`,
```
Comment (digitalsadhu) [inline lib/sink.js]: s/exist/exists

File: lib/sink.js [hunk 8]
```diff
@@ -73,61 +71,95 @@ class WriteStream extends stream.PassThrough {
...
+                return this.bucket;
+            });
+    }
+
+    _getGcsFile(fileName) {
+        return this.gcs.bucket(this.bucketName).file(fileName);
+    }
+
+    async get(fileName) {
+        const data = await this._getGcsFile(fileName).download();
```
Comment (digitalsadhu) [inline lib/sink.js]: should we be try/catching this and wrapping the error or are we happy with whatever error GCS might return reaching the consumer?

File: test/sink.test.js [hunk 9]
```diff
@@ -42,3 +45,113 @@ test('.writer() - no value for "type" argument - should throw', () => {
         sink.writer();
     }).toThrowError('"type" is missing');
 });
+
+test.only('.writer() - happy path', async done => {
```
Comment (digitalsadhu) [inline test/sink.test.js]: don't think you meant to leave the `.only` in here!

Comment (digitalsadhu) [reply test/sink.test.js]: were there no node warnings about returning a promise and calling a callback when using `async done =>` ?

File: test/sink.test.js [hunk 10]
```diff
@@ -42,3 +45,113 @@ test('.writer() - no value for "type" argument - should throw', () => {
...
+    source.pipe(dest);
+
+    dest.on('file not saved', () => {
+        done(new Error('File not saved'));
+    });
+
+    dest.on('file saved', (id, file) => {
+        expect(id).toMatchSnapshot();
+        expect(file).toMatchSnapshot();
+        process.nextTick(done);
```
Comment (digitalsadhu) [inline test/sink.test.js]: I think i've seen this somewhere else as well, wondering why do we need to do this?

File: test/sink.test.js [hunk 11]
```diff
@@ -42,3 +45,113 @@ test('.writer() - no value for "type" argument - should throw', () => {
...
+    });
+});
+
+test('.get() - should return null when ', async () => {
```
Comment (digitalsadhu) [inline test/sink.test.js]: when what? rasta copy pasta? 😁

File: test/sink.test.js [hunk 12]
```diff
@@ -42,3 +45,113 @@ test('.writer() - no value for "type" argument - should throw', () => {
...
+    const sink = getValidSink();
+
+    await sink.ready;
+
+    sink.gcs._setState({ save: '' });
+
+    const result = await sink.set('some-file', 'file-content');
+
+    expect(result).toBe(true);
+});
```
Comment (digitalsadhu) [inline test/sink.test.js]: This test is identical to the test above. This won't help you get more test coverage 😁

File: lib/sink.js [hunk 13]
```diff
@@ -73,61 +71,95 @@ class WriteStream extends stream.PassThrough {
...
+        this.bucketName = bucketName;
         this.name = 'asset-pipe-sink-gcs';
         this.gcs = gcsClient(options);
         this.bucket = null;
 
-        this.gcs.createBucket(bucket, metadata, (error, bucketObj) => {
-            if (error) {
-                this.bucket = this.gcs.bucket(bucket);
-                return this.emit(
+        this.ready = this.gcs
```
Comment (digitalsadhu) [inline lib/sink.js]: Is this intended as an external api change? If it is should this be documented in the readme? 
Do the other sinks also have a `.ready`? If not, maybe we should be being consistent with it?

File: test/sink.test.js (hunk 9 above)
Comment (sveisvei) [reply test/sink.test.js]: thx. I dont think async and done usage is exclusive.

File: test/sink.test.js (hunk 11 above)
Comment (sveisvei) [reply test/sink.test.js]: jjeesh, fixing

File: test/sink.test.js (hunk 12 above)
Comment (sveisvei) [reply test/sink.test.js]: delirious here...

File: lib/sink.js (hunk 8 above)
Comment (sveisvei) [reply lib/sink.js]: Unsure, if something is wrong - what kind of errors should we suppress?

File: lib/sink.js (hunk 13 above)
Comment (sveisvei) [reply lib/sink.js]: Thx - Yeah, this needs to be documented. Looks like in the code that we need a ready call....

File: README.md (hunk 3 above)
Comment (sveisvei) [reply README.md]: fixing.

File: lib/sink.js (hunk 13 above)
Comment (sveisvei) [reply lib/sink.js]: I could make it "private" for now? If we want to delay this.

Comment (digitalsadhu) [reply lib/sink.js]: how would you make it private? move the async work into the individual methods?

Comment (sveisvei) [reply lib/sink.js]: this._ready, as in javascript "'"private"'".

File: lib/sink.js (hunk 8 above)
Comment (digitalsadhu) [reply lib/sink.js]: Not sure what I was thinking, maybe some kind of wrapping like:

`'sink.get' was unable to retrieve file '${fileName}' from storage. Additionally got error: ${error}`

or some such. Might be a bit overkill?

File: lib/sink.js (hunk 13 above)
Comment (digitalsadhu) [reply lib/sink.js]: Yeah that might be a nice way to go.

Comment (sveisvei) [reply lib/sink.js]: although, creating a bucket as part of the startup might be a bad thing anyway - as this can be done in the UI/config.

Comment (sveisvei) [reply lib/sink.js]: updated with _ready as internal.

File: README.md [hunk 14]
```diff
@@ -76,70 +76,87 @@ This module have the following API:
...
 Events:
 
- - `storage info` - General info on the underlaying storage. Emits with: `message` and `metaObject`.
+* `storage info` - General info on the underlaying storage. Emits with:
+  `message` and `metaObject`.
 
+### get(fileName: string): Promise
 
-### writer(type)
+<string>
```
Comment (sveisvei) [inline README.md]: keeps newling this.... prettier...

Comment (SimenB) [reply README.md]: Update prettier to 1.8.2

**Pull request merged**